HTTP file server

elandau opened the pull request:
Initial implementation for serving static content from RxNetty based on https://github.com/netty/netty/blob/master/example/src/main/java/io/netty/example/http/file/HttpStaticFileServerHandler.java.  

Additional features include,
1.  Support for single and multiple file sources
2.  zero-copy serving of files

elandau committed:
First pass at an http file server with support for cache headers and multiple file sources

gorzell commented:
yay!

tbak commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/FileRequestHandler.java:
```diff
@@ -0,0 +1,100 @@
+                return FileResponses.sendError(response, HttpResponseStatus.NOT_FOUND);
+            }
+            
+            File file = new File(resource.toURI());
```
What if the file found on classpath is in jar not on the file system?

elandau replied:
This will handle files in jars as well

tbak commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/FileRequestHandler.java:
```diff
@@ -0,0 +1,100 @@
+        }
+        
+        try {
+            final String uri = request.getUri();
```
Should not we have UriResolver? URL class is broken, and we anyway convert it to URI here.

elandau replied:
I'll make the change.

tbak commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/OwnableRequestHandler.java:
```diff
@@ -0,0 +1,38 @@
+ * @param <O>
+ */
+public class OwnableRequestHandler<I, O> implements RequestHandler<I, O> {
+
```
This is useful not only for static content handling, so we could put it into server package.
And maybe find a better name, like RequestHandlerRouter or RequestHandlerDispatcher? 

Alternative implementation could follow the chain of responsibility pattern. Would be more generic, and applicable to other cases as well.

elandau replied:
+1 
A more generic solution would be great

NiteshKant replied:
Are we talking about a simple URI router much like discussed in issue #31 ?

If yes, I see it much like karyon's [SimpleUriRouter](https://github.com/Netflix/karyon/blob/master/karyon-core/src/main/java/com/netflix/karyon/transport/http/SimpleUriRouter.java)

I am for creating this utility to compose multiple `RequestHandler` as handling multiple URIs.

elandau replied:
I think this use case is different from what's implemented in SimpleUriRouter.  In SimpleUriRouter each uri prefix of regex is mapped to exactly one RequestHandler.  Here we can specify a cascading set of file sources that are inspected in a predefined order.  So, depending on which source contains the file the same URI may be end up using a different RequestHandler.  It would be equivalent to SimpleUriRouter mapping a URI to multiple RequestHandlers.

NiteshKant replied:
Sure but the URI -> `RequestHandler` mapping can also map to a composed `RequestHandler`, correct?

To concretize what I am saying:

``` java
public class SimpleUriRouter<I, O> implements RequestHandler<I, O> {

    public SimpleUriRouter<I, O> withRoute(String uri, RequestHandler<I, O> route) {
        // store the route in a linked hashmap to guarantee route matching in the order of addition
        return this;
    }

    @Override
    public Observable<Void> handle(HttpServerRequest<I> request, HttpServerResponse<O> response) {
        return Observable.empty(); //TODO: Run through the routes and choose whichever matches first
    }

    public static void main(String[] args) {
        SimpleUriRouter<ByteBuf, ByteBuf> router = new SimpleUriRouter<>();
        router.withRoute("/a", (request, response) -> Observable.empty());

        SimpleUriRouter<ByteBuf, ByteBuf> complexRoute = new SimpleUriRouter<>(); 
        complexRoute.withRoute("/b/c", (request, response) -> Observable.empty());
        complexRoute.withRoute("/b/d", (request, response) -> Observable.empty());

        router.withRoute("/b", complexRoute);
    }
}
```

tbak replied:
If we replace first argument in `withRoute` with a predicate function, we will get more flexibility. We could provide a few standard predicates for string prefix, regexp.

elandau replied:
But the above example still has a 1-1 relationship between path and handler.  The functionality I've implemented would be more along the lines of 

``` java
router.withRoute("/b", ImmutableList.of(new FileSystemBasedRouter("/var/files"), new ClassPathBaseRouter(), new RemoteUrlBasedRouter("http://someotherfileserver"));
```

The expected behavior would be that for any uri starting with "/b" we'd first look for the file in the folder /var/files.  If not found then look for the file in the class path.  If still not found then look for the file on a remote server.

gorzell replied:
I think this might be another place where it would be more clear if we had a good mechanism for building chains of handlers.  I think that is in essence what @elandau is trying to setup here.  But what each handler in the chain does with the request should be arbitrary.

This kind of gets back to the question of what is the difference between a chain of handlers and a pipeline and how we make both easy to configure and combine.

tbak commented on rx-netty/src/main/resources/META-INF/mime.types:
```diff
@@ -0,0 +1,1588 @@
+#
+# MIME type (lowercased)			Extensions
+# ============================================	==========
+# application/1d-interleaved-parityfec
```
Is there a reason that we have some mime types filtered out?
If yes, why not remove them from this file.

elandau replied:
Good question.  We may want to seriously trim the supported types in this file and make sure there is a mechanism to provide addition mime types.

NiteshKant commented on rx-netty/src/main/java/io/reactivex/netty/channel/DefaultChannelWriter.java:
```diff
@@ -102,6 +106,28 @@ public void writeString(String msg) {
+    
+    @Override
+    public Observable<Void> writeFileRegion(FileRegion region) {
+        final ChannelFuture future = writeOnChannel(region);
```
We should not use writeOnChannel() here as that adds the future of write to unflushed futures. This write does not need a flush.

Instead you can just use 

``` java
ChannelFuture writeFuture = getChannel().write(region);
```

elandau replied:
This doesn't work.  It seems like the write does need to be flushed after all.

NiteshKant commented on rx-netty/src/main/java/io/reactivex/netty/channel/ChannelWriter.java:
```diff
@@ -36,6 +37,8 @@
 
     void writeString(String msg);
 
+    Observable<Void> writeFileRegion(FileRegion region);
```
Should we instead return `Observable<Long>` and underneath use a `ProgressivePromise` to provide the status of the write in terms of bytes written?

The code then would be:

``` java
        final ChannelProgressivePromise progressivePromise = ctx.newProgressivePromise();
        getChannel().write(region, progressivePromise);
        return Observable.create(new Observable.OnSubscribe<Long>() {
            @Override
            public void call(final Subscriber<? super Long> subscriber) {
                progressivePromise.addListener(new ChannelProgressiveFutureListener() {
                    @Override
                    public void operationProgressed(ChannelProgressiveFuture future, long progress, long total)
                            throws Exception {
                        subscriber.onNext(progress); // sending progress, total can be calculated by the subscriber.
                    }

                    @Override
                    public void operationComplete(ChannelProgressiveFuture future) throws Exception {
                        if (future.isSuccess()) {
                            subscriber.onCompleted();
                        } else {
                            subscriber.onError(future.cause());
                        }
                    }
                });
            }
        });
```

elandau replied:
Makes sense but how will this Observable<Long> be consumed?

NiteshKant commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/FileRequestHandler.java:
```diff
@@ -0,0 +1,100 @@
+ *
+ */
+public class FileRequestHandler implements RequestHandler<ByteBuf, ByteBuf> {
+    public static final String IF_MODIFIED_SINCE = "If-Modified-Since";
```
You can use netty's HTTP header name: `io.netty.handler.codec.http.HttpHeaders.Names.IF_MODIFIED_SINCE`

NiteshKant commented on rx-netty/src/main/resources/META-INF/mime.types:
```diff
@@ -0,0 +1,1588 @@
+# This file maps Internet media types to unique file extension(s).
```
Where is this file used?

elandau replied:
This is used in FileResponses.setContentTypeHeader().

``` java
    public static void setContentTypeHeader(HttpServerResponse<ByteBuf> response, File file) {
        MimetypesFileTypeMap mimeTypesMap = new MimetypesFileTypeMap();
        response.getHeaders().set(HttpHeaders.CONTENT_TYPE, mimeTypesMap.getContentType(file.getPath()));
    }
```

See documentation here, http://docs.oracle.com/javase/7/docs/api/javax/activation/MimetypesFileTypeMap.html

elandau committed:
Use URI instead of URL
Don't auto-flush writeFileRegion

gorzell commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/ClassPathURLResolver.java:
```diff
@@ -0,0 +1,28 @@
+    }
+     
+    @Override
+    public URL getUrl(String path) {
```
Might be worth comment that this can return null.

gorzell commented:
No unit tests at all?!?

gorzell commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/ClassPathURIResolver.java:
```diff
@@ -0,0 +1,37 @@
+ *
+ */
+public class ClassPathURIResolver implements URIResolver {
+    private static final String DEFAULT_PATH_PREFIX = "WEB-INF";
```
Should you check for and enforce that this ends with a "/"?  You add a file separator in the LocalDirectoryResolver.

elandau replied:
done

gorzell commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/LocalDirectoryURIResolver.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    public URI getUri(String path) {
+        try {
+            return new URI("file:///" + prefix + path);
```
There is a security issue here, there is nothing to prevent the caller from navigating the filesystem.  This may also exist for the classpath, but I am having a harder time proving that.

```
val uri2 = new URI(s"file:///$prefix/../../etc/hosts")
uri2: java.net.URI = file:////Users/gorzell/../../etc/hosts

scala> import java.io.File
import java.io.File

scala> new File(uri2).exists
res10: Boolean = true
```

Some unit tests with these kinds of cases would really help.

elandau replied:
done

gorzell commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/FileRequestHandler.java:
```diff
@@ -0,0 +1,98 @@
+            FileResponses.setDateAndCacheHeaders(response, file);
+            response.writeFileRegion(new DefaultFileRegion(raf.getChannel(), 0, fileLength));
+            return response.close();
+//            if (HttpHeaderUtil.isKeepAlive(request)) {
```
Probably worth cleaning this up to avoid confusion.

elandau replied:
done

gorzell commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/URIResolver.java:
```diff
@@ -0,0 +1,18 @@
+     * Resolve the path URI to a local URL 
+     * 
+     * @param uri
+     * @return A local URL or null if not found
```
This isn't really accurate (and has a typo).  The LocalFile implementation doesn't check that the file is "found" or exists at all.  Short of a totally malformed string, not much in the way of sanity checks are promised.

elandau replied:
The contract should be null if not found.  I fixed LocalDirectoryURIResolver to check for existence.

gorzell commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/UrlUtils.java:
```diff
@@ -0,0 +1,38 @@
+import java.net.URLDecoder;
+import java.util.regex.Pattern;
+
+public class UrlUtils {
```
This is at least a start, I don't see where it is ever used though.  Is the find feature in Chrome failing me?

elandau replied:
Using this somehow got dropped from the original code.  It's back in there now.

gorzell commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/URIResolver.java:
```diff
@@ -0,0 +1,18 @@
+ * 
+ * @author elandau
+ */
+public interface URIResolver {
```
I'd suggest an abstract base class that does some of the basics like sanitizing the input (unless I missed where that is happening) which could then be extended by the two existing impls you have.

elandau replied:
Sanitize now happens early on so we can return a 403

elandau committed:
Enable basic guard against basic hacks
Add unit tests
Check for file existence when using LocalDirectoryURIResolver

elandau committed:
Clean up implementation to merge the URIResolver into specific subclasses of FileRequestHandler.
Remove support for cascading file sources.  This may be added later.
Introduce HttpError to attach an HTTP response code to exceptions

NiteshKant commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/LocalDirectoryRequestHandler.java:
```diff
@@ -0,0 +1,40 @@
+            }
+            return uri;
+        } catch (URISyntaxException e) {
+            return null;
```
Log the exception?

NiteshKant commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/WebappFileRequestHandler.java:
```diff
@@ -0,0 +1,7 @@
+package io.reactivex.netty.protocol.http.server.file;
+
+public class WebappFileRequestHandler extends ClassPathFileRequestHandler {
```
We don't really have much webapp support perse :)

NiteshKant commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/LocalDirectoryRequestHandler.java:
```diff
@@ -0,0 +1,40 @@
+    @Override
+    protected URI resolveUri(String path) {
+        try {
+            URI uri = new URI("file:///" + prefix + path);
```
May be you can pre-create `"file:///" + prefix` string to avoid one more string creation?

NiteshKant commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/LocalDirectoryRequestHandler.java:
```diff
@@ -0,0 +1,40 @@
+import java.io.File;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
```
Looks like this is a java 7 API, we are java 6 compatible.

NiteshKant commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/FileRequestHandler.java:
```diff
@@ -0,0 +1,111 @@
+    public Observable<Void> handle(HttpServerRequest<ByteBuf> request, HttpServerResponse<ByteBuf> response) {
+        // We don't support GET.  
+        if (!request.getHttpMethod().equals(HttpMethod.GET)) {
+            return Observable.error(new HttpError(HttpResponseStatus.NOT_FOUND));
```
If GET is not supported, then should the response code be `405` => Method Not Allowed?

BTW, why is GET not supported?

elandau replied:
Only GET is supported.  I'll change the response to 405.

NiteshKant commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/FileRequestHandler.java:
```diff
@@ -0,0 +1,111 @@
+                long fileLastModifiedSeconds = file.lastModified() / 1000;
+                if (ifModifiedSinceDateSeconds == fileLastModifiedSeconds) {
+                    response.setStatus(HttpResponseStatus.NOT_MODIFIED);
+                    setDateHeader(response);
```
Is there a value in passing the `SimpleDateFormat` instance to this method? This will also aid method reuse from `setDateAndCacheHeaders`

NiteshKant commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/FileRequestHandler.java:
```diff
@@ -0,0 +1,111 @@
+            // TODO: Handle keep alive headers
+            return response.close().finallyDo(closeFileAction(raf));
+        }
+        catch (Exception e) {
```
You do not have to add this error handling logic as `HttpServer` does the handling like this by default.

elandau replied:
Hmmm... but I have to catch the checked exceptions somewhere.

NiteshKant replied:
Oh there are checked exceptions thrown by this code? Since, it was effectively the entire method in a try-catch and rxnetty APIs not throwing an exception I thought it is just a safe-guard.

NiteshKant commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/FileRequestHandler.java:
```diff
@@ -0,0 +1,111 @@
+            setDateAndCacheHeaders(response, file);
+            response.writeFileRegion(new DefaultFileRegion(raf.getChannel(), 0, fileLength));
+            
+            // TODO: Handle keep alive headers
```
Is this a TODO we can't resolve now?

NiteshKant commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/FileRequestHandler.java:
```diff
@@ -0,0 +1,111 @@
+            response.getHeaders().setContentLength(fileLength);
+            setContentTypeHeader(response, file);
+            setDateAndCacheHeaders(response, file);
+            response.writeFileRegion(new DefaultFileRegion(raf.getChannel(), 0, fileLength));
```
The [netty example](https://github.com/netty/netty/blob/master/example/src/main/java/io/netty/example/file/FileServerHandler.java) suggests that we can not use this for HTTPS. Should we be handling SSL case?

NiteshKant commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/file/FileRequestHandler.java:
```diff
@@ -0,0 +1,111 @@
+            response.writeFileRegion(new DefaultFileRegion(raf.getChannel(), 0, fileLength));
+            
+            // TODO: Handle keep alive headers
+            return response.close().finallyDo(closeFileAction(raf));
```
By reading code it appears that netty will close the file after writing it on the channel. 

Specifically, the method [DefaultFileRegion.deallocate()](https://github.com/netty/netty/blob/master/example/src/main/java/io/netty/example/file/FileServerHandler.java) will be called by [ChannelOutboundBuffer](https://github.com/netty/netty/blob/master/transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java#L246) post transfer. So, I think we can safely remove this close handler.

NiteshKant commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/RequestHandlerWithErrorMapper.java:
```diff
@@ -0,0 +1,46 @@
+ * @param <O>
+ */
+public class RequestHandlerWithErrorMapper<I, O> implements RequestHandler<I, O> {
+    private Func1<Throwable, ErrorResponseGenerator<O>> errorMapper;
```
Can this just be `Func2<Throwable, HttpServerResponse<I>, Observable<Void>>` or just `Func2<Throwable, HttpServerResponse<I>, Void>`?

elandau replied:
Don't we want to promote reuse of ErrorResponseGenerator as the contract for generating error responses?  Why introduce another contract.

NiteshKant commented on rx-netty/src/main/java/io/reactivex/netty/protocol/http/server/RequestHandlerWithErrorMapper.java:
```diff
@@ -0,0 +1,46 @@
+    @Override
+    public Observable<Void> handle(HttpServerRequest<I> request, final HttpServerResponse<O> response) {
+        return handler.handle(request, response)
+                .onErrorResumeNext(new Func1<Throwable, Observable<Void>>() {
```
Does it make sense to just do a `doOnError()` so that you can avoid creating and managing an `Observable`. Specifically since the `errorMapper` are sync (returns void), it should just be fine?

elandau replied:
Won't doOnError still propagate the error to to the subscriber of Observable<Void>.  The idea here is to write the error response and not propagate the error to the container.

elandau committed:
Update from code review

elandau committed:
Fix typos

elandau committed:
Cleanup checked exception handling
Add more logging
Support for keep alive headers
Support for chunked response
Support for SSL
